packages/api/pages/api/users/[userId].ts

```diff
@@ -3,20 +3,27 @@ import { SystemRole, UpdateUserRequestBody } from '@translation/api-types';
 import { authorize } from '../../../shared/access-control/authorize';
 import createRoute from '../../../shared/Route';
 import { client } from '../../../shared/db';
+import { auth } from '../../../shared/auth';
+import { randomBytes } from 'crypto';
+import mailer from '../../../shared/mailer';
+import { redirects } from '../../../shared/redirects';
 
 export default createRoute<{ userId: string }>()
   .patch<UpdateUserRequestBody, void>({
     schema: z.object({
+      name: z.string().optional(),
+      password: z.string().optional(),
+      email: z.string().optional(),
       systemRoles: z
         .array(
           z.enum(Object.values(SystemRole) as [SystemRole, ...SystemRole[]])
         )
         .optional(),
     }),
-    authorize: authorize({
-      action: 'administer',
+    authorize: authorize((req) => ({
+      action: req.body.systemRoles ? 'administer' : 'update',
```

Comment:
Is there ever the possibility that `req.body.systemRoles` could be the empty array? In that situation, I think your condition might not work as intended.

```suggestion
      action: req.body.systemRoles != undefined ? 'administer' : 'update',
```

Reply:
In that scenario, it means that an admin has removed all roles from a user. Right now there is only one system role of admin, so to remove it, you have to pass an empty array.

Reply:
So when an empty array is passed, we still want the action to be `administer`, correct? The logic as written would evaluate to `action: 'update'` when the `systemRoles` is empty.

Reply:
Ok I was mistaken. I thought the empty array evaluates to `false` in a boolean context.